### PR TITLE
[SettingsBundle] Add API support to SettingsController

### PR DESCRIPTION
**Pull request by aramalipoor**
Q  | A
------------- | -------------
Bug fix?  | no
New feature?  | yes
BC breaks? | no
Deprecations? | no
Fixed tickets | -
License | MIT
Doc PR | -

### CODE REVIEW

**Comment by aramalipoor** [comment]
@pjedrzejewski Travis is all green; there's just one unrelated fail!

**Comment by aramalipoor** [comment]
@pjedrzejewski @aRn0D ping!

File: src/Sylius/Bundle/SettingsBundle/Controller/SettingsController.php [hunk 1]
```diff
@@ -56,16 +54,21 @@ public function updateAction(Request $request, $namespace)
                 $message = $this->getTranslator()->trans($exception->getMessage(), array(), 'validators');
                 $messageType = 'error';
             }
-            $request->getSession()->getBag('flashes')->add($messageType, $message);
 
-            if ($request->headers->has('referer')) {
-                return $this->redirect($request->headers->get('referer'));
+            if ($this->isApiRequest()) {
+                return $this->handleView($this->view($settings, 204));
+            } else {
```
Comment (stloyd) [inline src/Sylius/Bundle/SettingsBundle/Controller/SettingsController.php]: `else` is useless as you call `return` in `if`.

File: src/Sylius/Bundle/SettingsBundle/Controller/SettingsController.php [hunk 2]
```diff
@@ -118,4 +121,9 @@ protected function isGrantedOr403($namespace)
 
         return false;
     }
+
+    public function isApiRequest()
+    {
+        return 'html' !== $this->getRequest()->getRequestFormat();
```
Comment (stloyd) [inline src/Sylius/Bundle/SettingsBundle/Controller/SettingsController.php]: Please don't use `getRequest()`, pass the Request as variable instead.

File: src/Sylius/Bundle/SettingsBundle/Form/Factory/SettingsFormFactory.php [hunk 3]
```diff
@@ -50,10 +50,12 @@ public function __construct(SchemaRegistryInterface $schemaRegistry, FormFactory
     /**
      * {@inheritdoc}
      */
-    public function create($namespace)
+    public function create($namespace, $data = null, $options = array())
     {
         $schema = $this->schemaRegistry->getSchema($namespace);
-        $builder = $this->formFactory->createBuilder('form', null, array('data_class' => null));
+        $builder = $this->formFactory->createBuilder('form', $data, array_merge_recursive(
+            $options, array('data_class' => null)
```
Comment (stloyd) [inline src/Sylius/Bundle/SettingsBundle/Form/Factory/SettingsFormFactory.php]: Should be opposide, `$options` have higher priority than defaults.

File: src/Sylius/Bundle/SettingsBundle/Form/Factory/SettingsFormFactory.php [hunk 4]
```diff
@@ -50,10 +50,12 @@ public function __construct(SchemaRegistryInterface $schemaRegistry, FormFactory
     /**
      * {@inheritdoc}
      */
-    public function create($namespace)
+    public function create($namespace, $data = null, $options = array())
```
Comment (stloyd) [inline src/Sylius/Bundle/SettingsBundle/Form/Factory/SettingsFormFactory.php]: `$options` should be type-hinted to array.

**Comment by aramalipoor** [comment]
@stloyd Thank you for your feedback :)

File: src/Sylius/Bundle/SettingsBundle/Controller/SettingsController.php [hunk 5]
```diff
@@ -118,4 +122,13 @@ protected function isGrantedOr403($namespace)
 
         return false;
     }
+
+    /**
+     * @param Request $request
+     * @return bool
+     */
+    public function isApiRequest($request)
```
Comment (stloyd) [inline src/Sylius/Bundle/SettingsBundle/Controller/SettingsController.php]: I guess this can be private, and you should type-hint Request object.

Comment (aramalipoor) [inline src/Sylius/Bundle/SettingsBundle/Controller/SettingsController.php]: Done Sir! :)

**Comment by aramalipoor** [comment]
@stloyd Thank you for your feedback :)

**Comment by aramalipoor** [comment]
@pjedrzejewski ping!

**Comment by aramalipoor** [comment]
@pjedrzejewski @aRn0D ping!

**Comment by aramalipoor** [comment]
@pjedrzejewski ping!

**Comment by pjedrzejewski** [comment]
Thank you Aram! :+1:

**Pull request merged**